### PR TITLE
BUG: Coerce to native types to_dict list and index (#46751)

### LINKED ISSUE #46751: BUG: to_dict("index") and to_dict("list") don't coerce to native types
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import numpy as np
import pandas as pd


def main():
    i64 = np.int64
    f64 = np.float64
    df = pd.DataFrame({
        "A": [i64(1), "ASD", f64(2.2)],
        "B": [(2,), f64(2.2), i64(1)],
    })
    a = df.to_dict("list")
    print("to_dict('list')")
    for key, list_ in a.items():
        for value in list_:
            print(f"{key}: {value}: {type(value)}")
    b = df.to_dict("index")
    print("to_dict('index')")
    for index, obj in b.items():
        for key, value in obj.items():
            print(f"{key}: {value}: {type(value)}")


if __name__ == "__main__":
    main()
```

### Issue Description

When using to_dict ONLY for orient=list/index the values are not coerced to python native types. We are simply missing a `maybe_box_native` call

### Expected Behavior

Each value in the resulting `to_dict` output should be Python native types, instead we are getting int64's and float64's.

This ONLY happens for orient=list and orient=index, the other orients correctly coerce to native types.

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit           : 4bfe3d07b4858144c219b9346329027024102ab6
python           : 3.9.10.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 20.6.0
Version          : Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu-7195.141.6~3/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.4.2
numpy            : 1.22.1
pytz             : 2021.1
dateutil         : 2.8.1
pip              : 22.0.3
setuptools       : 57.0.0
Cython           : 0.29.23
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : 1.3.7
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.9.1
jinja2           : 3.0.1
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : 0.6.2
gcsfs            : None
markupsafe       : 2.0.1
matplotlib       : 3.4.2
numba            : 0.53.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 4.0.1
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.6.3
snappy           : None
sqlalchemy       : 1.1.12
tables           : None
tabulate         : None
xarray           : None
xlrd             : 1.1.0
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by RogerThomas**
- [x] closes #46751 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: doc/source/whatsnew/v1.5.0.rst [hunk 1]
```diff
@@ -500,6 +500,7 @@ Conversion
 - Bug in :func:`array` with ``FloatingDtype`` and values containing float-castable strings incorrectly raising (:issue:`45424`)
 - Bug when comparing string and datetime64ns objects causing ``OverflowError`` exception. (:issue:`45506`)
 - Bug in metaclass of generic abstract dtypes causing :meth:`DataFrame.apply` and :meth:`Series.apply` to raise for the built-in function ``type`` (:issue:`46684`)
+- Bug in :meth:`DataFrame.to_dict` for ``orient=="list"`` or ``orient=="index"`` was not returning native types. (:issue:`46751`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v1.5.0.rst:503]: nit: specify as the user would when calling the method; e.g. `orient="list"`. Also no period before the issue #.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Looks great - very nice test. Small nits on the whatsnew.

File: doc/source/whatsnew/v1.5.0.rst (hunk 1 above)
Comment (RogerThomas) [reply doc/source/whatsnew/v1.5.0.rst]: @rhshadrach done

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by RogerThomas** [comment]
@rhshadrach looks like there are 2 tests failing that don't seem in any way related to my code, how should I proceed?

**Comment by rhshadrach** [comment]
It's actually just one; the `pandas-dev.pandas` will fail anytime one of the Azure pipeline build fails, so it's just the py38_32bit. This build is failing on all PRs for reasons I haven't been able to figure out; okay to ignore.

**Comment by rhshadrach** [comment]
Thanks @RogerThomas!

**Pull request merged**